src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,305 @@
+
+For example, we can use the property access as a parameter.
+
+```javascript
+console.log(o.foo);
+```
+
+In this case, the behavior is defined in the runtime semantics of `ArgumentList` production which calls `GetValue` on the argument:
+
+> [Runtime Semantics: ArgumentListEvaluation](https://tc39.es/ecma262/#sec-argument-lists-runtime-semantics-argumentlistevaluation)
+>
+> `ArgumentList : AssignmentExpression`
+>
+> 1. Let `ref` be the result of evaluating `AssignmentExpression`.
+> 2. Let `arg` be `? GetValue(ref)`.
+> 3. Return a List whose sole item is `arg`.
+
+Now the `AssignmentExpression` is `o.foo` and the result of evaluating it is the above mentioned Reference. Now we call `GetValue` on it.
+
+### Property access as the right hand side of an assignemnt
+
+We can also use the property access as a right hand side of an assignment:
+
+```javascript
+const x = o.foo;
+```
+
+In this case, the behavior is defined in the runtime semantics for the `AssignmentExpression` production which calls `GetValue` on the right hand side:
+
+> [Runtime Semantics: Evaluation for `AssignmentExpression : LeftHandSideExpression = AssignmentExpression`](https://tc39.es/ecma262/#sec-assignment-operators-runtime-semantics-evaluation)
+>
+> 1. If `LeftHandSideExpression` is neither an `ObjectLiteral` nor an `ArrayLiteral`, then
+> a. Let `lref` be the result of evaluating `LeftHandSideExpression`.
+> b. ReturnIfAbrupt(lref).
+> c. If `IsAnonymousFunctionDefinition(AssignmentExpression)` and `IsIdentifierRef` of `LeftHandSideExpression` are both `true`, then
+> i. Let `rval` be `NamedEvaluation` of `AssignmentExpression` with argument `GetReferencedName(lref)`.
+> d. Else,
+> i. Let `rref` be the result of evaluating `AssignmentExpression`.
+> ii. Let `rval` be `? GetValue(rref)`.
+> e. Perform `? PutValue(lref, rval)`.
+> f. Return `rval`.
+> 2. Let `assignmentPattern` be the `AssignmentPattern` that is covered by `LeftHandSideExpression`.
+> 3. Let `rref` be the result of evaluating `AssignmentExpression`.
+> 4. Let `rval` be `? GetValue(rref)`.
+> 5. Perform `? DestructuringAssignmentEvaluation` of `assignmentPattern` using `rval` as the argument.
+> 6. Return `rval`.
+
+Now the `LeftHandSideExpression` is not an object literal or an array literal, so we take the if branch in step 1. The assignment expression is `o.foo`, and it's not a function definition, so we take the else branch in 1 d. There we evaluate the `AssignmentExpression` and call `GetValue` on it.
+
+In any case, `GetValue` will be called on the Reference which is the result of evaluating the `MemberExpression`. Thus, we know that the Object internal method `[[Get]]` will get invoked when accessing a property on an Object, and the prototype chain walk will occur.
+
+### Why is `o2.foo` an `AssignmentExpression`?
+
+Just one more thing. Previously, we used `o2.foo` as an `AssignmentExpression`. It doesn't look like one though. Why is it an `AssignmentExpression`?
```

Comment:
What does this "Previously" refer to? Isn't o2.foo used the last time before `AssignmentExpression` is even defined?

Reply:
Oops, this was supposed to say o.foo. However, I fixed the examples so that all access o2.foo instead of o.foo, so it's more consistent (the blog post talks about accessing o2.foo throughout).

Reply:
Also added a forward-reference to the place where I say o2.foo is an AssignmentExpression.